main/src/ca/uwaterloo/flix/language/phase/Enumgen.scala

```diff
@@ -0,0 +1,353 @@
+package ca.uwaterloo.flix.language.phase
+
+import org.objectweb.asm
+import org.objectweb.asm.{ClassWriter, Label}
+import org.objectweb.asm.Opcodes._
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.InternalCompilerException
+
+/**
+  * Created by ramin on 2017-05-22.
+  *
+  * Generates classes and interfaces which will represent enums
+  */
+object Enumgen {
+  import CodegenHelper._
+
+  /**
+    * Generates an interface for TagInterface which includes one method which
+    * returns the name of the enum case which implements this interface
+    * @param name name of the class
+    * @return byte code representation of the class
+    */
+  def compileTagInterface(name: List[String]) : Array[Byte] = {
```

Comment:
I much prefer a case class like `QualName` since that at least carries some semantic information. Perhaps I might suggest classes like:

`JavaClassName`, `JavaInterfaceName`, etc. etc. if that makes sense?

I understand that some of the older code uses strings - but I am not satisfied with that either :)